ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/SyncTargetManager.java

```diff
@@ -0,0 +1,123 @@
+import tech.pegasys.pantheon.ethereum.mainnet.ProtocolSchedule;
+import tech.pegasys.pantheon.metrics.LabelledMetric;
+import tech.pegasys.pantheon.metrics.OperationTimer;
+
+import java.time.Duration;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public abstract class SyncTargetManager<C> {
+
+  private static final Logger LOG = LogManager.getLogger();
+  private volatile long syncTargetDisconnectListenerId;
+  private volatile boolean syncTargetDisconnected = false;
+  private final SynchronizerConfiguration config;
+  private final ProtocolSchedule<C> protocolSchedule;
+  private final ProtocolContext<C> protocolContext;
+  private final EthContext ethContext;
+  private final SyncState syncState;
+  private final LabelledMetric<OperationTimer> ethTasksTimer;
+
+  public SyncTargetManager(
+      final SynchronizerConfiguration config,
+      final ProtocolSchedule<C> protocolSchedule,
+      final ProtocolContext<C> protocolContext,
+      final EthContext ethContext,
+      final SyncState syncState,
+      final LabelledMetric<OperationTimer> ethTasksTimer) {
+    this.config = config;
+    this.protocolSchedule = protocolSchedule;
+    this.protocolContext = protocolContext;
+    this.ethContext = ethContext;
+    this.syncState = syncState;
+    this.ethTasksTimer = ethTasksTimer;
+  }
+
+  public CompletableFuture<SyncTarget> findSyncTarget() {
+    final Optional<SyncTarget> maybeSyncTarget = syncState.syncTarget();
+    if (maybeSyncTarget.isPresent()) {
+      // Nothing to do
+      return CompletableFuture.completedFuture(maybeSyncTarget.get());
+    }
+
+    final Optional<EthPeer> maybeBestPeer = selectBestAvailableSyncTarget();
+    if (maybeBestPeer.isPresent()) {
+      final EthPeer bestPeer = maybeBestPeer.get();
+      return DetermineCommonAncestorTask.create(
+              protocolSchedule,
+              protocolContext,
+              ethContext,
+              bestPeer,
+              config.downloaderHeaderRequestSize(),
+              ethTasksTimer)
+          .run()
+          .handle((r, t) -> r)
+          .thenCompose(
+              (target) -> {
+                if (target == null) {
+                  return waitForPeerAndThenSetSyncTarget();
+                }
+                final SyncTarget syncTarget = syncState.setSyncTarget(bestPeer, target);
+                LOG.info(
+                    "Found common ancestor with peer {} at block {}", bestPeer, target.getNumber());
+                syncTargetDisconnectListenerId =
+                    bestPeer.subscribeDisconnect(this::onSyncTargetPeerDisconnect);
+                return CompletableFuture.completedFuture(syncTarget);
+              });
+    } else {
+      return waitForPeerAndThenSetSyncTarget();
+    }
+  }
+
+  protected abstract Optional<EthPeer> selectBestAvailableSyncTarget();
```

Comment:
This difference here for fast sync is an extra check that the sync target agrees the pivot block is on the canonical chain?

Reply:
Yep exactly.